### ISSUE TITLE
RAM keep increasing during inference using DefaultPredictor

### ISSUE THREAD

**Issue by ArtificialNotImbecile** [facebookresearch/detectron2#1728]
## Instructions To Reproduce the 🐛 Bug:

1. what changes you made (`git diff`) or what code you wrote
```
No changes
```
2. what exact command you run:
```
import cv2
from detectron2.config import get_cfg
from detectron2.engine import DefaultPredictor
import detectron2

cfg = get_cfg()
cfg_file = "path/to/detectron2/configs/COCO-Detection/retinanet_R_50_FPN_1x.yaml"
cfg.merge_from_file(cfg_file)
cfg.MODEL.RETINANET.SCORE_THRESH_TEST = 0.5
cfg.INPUT.MIN_SIZE_TEST = 0 
cfg.MODEL.WEIGHTS = 'path/to/retinanet1x_weight.pkl' #download from model zoo
model = DefaultPredictor(cfg)
if __name__ == "__main__":
    image_path = 'path/to/image.jpg'
    image = cv2.imread(image_path)
    for i in range(10000000):
        r = model(image)
```
3. what you observed (including __full logs__):
```
Memory usage keeps increasing(I use htop to observe memory usage from time to time) and finally eats up all available RAM
```
4. please simplify the steps as much as possible so they do not require additional resources to
	 run, such as a private dataset.

## Expected behavior:

I think memory should be fixed in a short time.

## Environment:

Provide your environment information using the following command:
```
----------------------  ---------------------------------------------------------
sys.platform            linux
Python                  3.6.9 (default, Nov  7 2019, 10:44:02) [GCC 8.3.0]
numpy                   1.16.4
detectron2              0.1.3 @/notebooks/detectron2/detectron2
Compiler                GCC 7.4
CUDA compiler           10.0
detectron2 arch flags   sm_75
DETECTRON2_ENV_MODULE   <not set>
PyTorch                 1.4.0 @/usr/local/lib/python3.6/dist-packages/torch
PyTorch debug build     False
CUDA available          True
GPU 0,1,2,3             Tesla T4
CUDA_HOME               /usr/local/cuda
Pillow                  6.2.1
torchvision             0.5.0 @/usr/local/lib/python3.6/dist-packages/torchvision
torchvision arch flags  sm_35, sm_50, sm_60, sm_70, sm_75
fvcore                  0.1.1.post200513
cv2                     4.1.1
----------------------  ---------------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - Intel(R) Math Kernel Library Version 2019.0.4 Product Build 20190411 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.21.1 (Git Hash 7d2fd500bc78936d1d648ca713b901012f470dbc)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.1
  - Build settings: BLAS=MKL, BUILD_NAMEDTENSOR=OFF, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Wno-stringop-overflow, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF, 
```

If your issue looks like an installation issue / environment issue,
please first try to solve it yourself with the instructions in
https://detectron2.readthedocs.io/tutorials/install.html#common-installation-issues

**Comment by ppwwyyxx** [comment]
please use the model following the documentation:https://detectron2.readthedocs.io/tutorials/models.html#use-a-model

**Comment by ArtificialNotImbecile** [comment]
> please use the model following the documentation:https://detectron2.readthedocs.io/tutorials/models.html#use-a-model
The code uses DefaultPredictor and doing nothing fancy. And using models like code below, as the documentation says, also leads to the same problem. I'll try to reproduce it with detectron2's docker environment.
```
import cv2
import torch
import detectron2
from detectron2.checkpoint import DetectionCheckpointer
from detectron2.config import get_cfg
from detectron2.modeling import build_model
import time
cfg = get_cfg()
cfg_file =  "path/to/detectron2/configs/COCO-Detection/retinanet_R_50_FPN_1x.yaml"
cfg.merge_from_file(cfg_file)
cfg.MODEL.RETINANET.SCORE_THRESH_TEST = 0.5
cfg.MODEL.WEIGHTS = '/path/to/retinanet1x_weight.pkl' # from model zoo
model = build_model(cfg)
model.eval()
checkpointer = DetectionCheckpointer(model)
checkpointer.load(cfg.MODEL.WEIGHTS)
with torch.no_grad():
    image_path = '/path/to/image.jpg'
    image = cv2.imread(image_path)
    height, width = image.shape[:2]
    image = torch.as_tensor(image.astype("float32").transpose(2, 0, 1))
    inputs = {"image": image, "height": height, "width": width}
    for i in range(10000000):
        predictions = model([inputs])[0] # this particular line cause RAM keep increasing
```

**Comment by ppwwyyxx** [comment]
> using models like code below

I observed no memory leak running this code. If you're able to reproduce it in the docker please let us know.

**Comment by ArtificialNotImbecile** [comment]
Steps to reproduce(I can only reproduce with torch==1.4.0, torchvisio==0.5.0(Both cuda 10.0 and cuda10.1), not with  torch==1.5 torchvision==0.6(cuda 10.1)):
1. Docker build and run as in docker/README, but without `--env="DISPLAY" --volume="/tmp/.X11-unix:/tmp/.X11-unix:rw"` since I'm not available to a display.

2. Install torch==1.4.0 and torchvision==0.5.0 with `pip install --user torch==1.4.0 torchvision==0.5.0 -f https://download.pytorch.org/whl/cu101/torch_stable.html`

3. Reinstall detectron2 `rm -rf  build && pip install  --user -e .`  # under /home/appuser/detectron2_repo

4. Run the following command(add  `cv2.resize(image, (1920, 1080)` since small image, say 224x224, seems not able, or slow to reproduce. As for 1920x1080 image, I observed approximately 1MB/Minute increase rate on my machine):
```
import cv2
import torch
import detectron2
from detectron2.checkpoint import DetectionCheckpointer
from detectron2.config import get_cfg
from detectron2.modeling import build_model
import time
cfg = get_cfg()
cfg_file =  "path/to/detectron2/configs/COCO-Detection/retinanet_R_50_FPN_1x.yaml"
cfg.merge_from_file(cfg_file)
cfg.MODEL.RETINANET.SCORE_THRESH_TEST = 0.5
cfg.MODEL.WEIGHTS = '/path/to/retinanet1x_weight.pkl' # from model zoo
model = build_model(cfg)
model.eval()
checkpointer = DetectionCheckpointer(model)
checkpointer.load(cfg.MODEL.WEIGHTS)
with torch.no_grad():
    image_path = '/path/to/image.jpg'
    image = cv2.imread(image_path)
    image = cv2.resize(image, (1920, 1080)) # make sure our image has the same size, 
    height, width = image.shape[:2]
    image = torch.as_tensor(image.astype("float32").transpose(2, 0, 1))
    inputs = {"image": image, "height": height, "width": width}
    for i in range(10000000):
        predictions = model([inputs])[0] # this particular line cause RAM keep increasing
```

**Comment by ppwwyyxx** [comment]
Thanks for the detailed investigation. I think the investigation suggests that the leak comes from pytorch, and since it seems already fixed in 1.5, there isn't much to do (pytorch doesn't do bugfix to previous versions AFAIK). There are similar reports like this, e.g. https://github.com/pytorch/pytorch/issues/38717

**Comment by ArtificialNotImbecile** [comment]
@ppwwyyxx  Thanks for the link. I upgrade pytorch to 1.5.0 and torchvision to 0.6.0 in my original environment and the problem is gone.
